File: pt5/1115.print-foo-bar-alternately.go
package main

type FooBar struct {
	n       int
	fooTurn chan struct{}
	barTurn chan struct{}
}

func Constructor(n int) FooBar {
	fooTurn := make(chan struct{}, 1)
	barTurn := make(chan struct{}, 1)
	fooTurn <- struct{}{}
	return FooBar{n: n, fooTurn: fooTurn, barTurn: barTurn}
}

func (fb *FooBar) foo(printFoo func()) {
	for i := 0; i < fb.n; i++ {
		<-fb.fooTurn
		printFoo()
		fb.barTurn <- struct{}{}
	}
}

func (fb *FooBar) bar(printBar func()) {
	for i := 0; i < fb.n; i++ {
		<-fb.barTurn
		printBar()
		fb.fooTurn <- struct{}{}
	}
}

func (fb *FooBar) Foo(printFoo func()) {
	fb.foo(printFoo)
}

func (fb *FooBar) Bar(printBar func()) {
	fb.bar(printBar)
}

File: pt5/1115.print-foo-bar-alternately.py
from __future__ import annotations

from threading import Semaphore
from typing import Callable


class FooBar:
    def __init__(self, n):
        self.n = n
        self.foo_turn = Semaphore(1)
        self.bar_turn = Semaphore(0)

    def foo(self, printFoo: Callable[[], None]) -> None:
        for _ in range(self.n):
            self.foo_turn.acquire()
            printFoo()
            self.bar_turn.release()

    def bar(self, printBar: Callable[[], None]) -> None:
        for _ in range(self.n):
            self.bar_turn.acquire()
            printBar()
            self.foo_turn.release()

